url71665d7d68_branscha_lib_scripty_tgz-pJ8-com_sdicons_scripty_ScriptyAutoReplJ8/src/com/sdicons/scripty/parser/EvalTrace.java

```diff
@@ -0,0 +1,921 @@
+            lBuilder.append(this.name).append(", ").append("and-composite").append(enabled?", enabled":", paused");
+            for(IBreakpoint lBp: bps)
+                lBuilder.append("\n\t * ").append(lBp.toString().replace("\n\t", "\n\t\t"));
+            return lBuilder.toString();
+        }
+    }
+
+    public static class BreakpointOr
+    implements IBreakpoint
+    {
+        private String name;
+        private boolean enabled = true;
+        private List<IBreakpoint> bps;
+
+        public BreakpointOr(String aName, List<IBreakpoint> aBps)
+        {
+            this.name = aName;
+            this.bps = aBps;
+        }
+
+        public String getName()
+        {
+            return name;
+        }
+
+        public void setEnabled(boolean enabled)
+        {
+            this.enabled = enabled;
+        }
+
+        public boolean breakHere(Eval2.EvalStack aStack)
+        {
+            if(!enabled) return false;
+            boolean lBreak = false;
+            for(IBreakpoint lBp: bps)
+            {
+                lBreak = lBreak || lBp.breakHere(aStack);
+                if(lBreak) return lBreak;
+            }
+            return lBreak;
+        }
+
+        @Override
+        public String toString()
+        {
+            StringBuilder lBuilder = new StringBuilder();
+            lBuilder.append(this.name).append(", ").append("or-composite").append(enabled?", enabled":", paused");
+            for(IBreakpoint lBp: bps)
+                lBuilder.append("\n\t * ").append(lBp.toString().replace("\n\t", "\n\t\t"));
+            return lBuilder.toString();
+        }
+    }
+
+    public EvalTrace(final Eval2 aEval, final Object aExpr)
+    {
+        // Keep track of the eval under scrutiny.
+        // Add an event listener to it.
+        eval = aEval;
+        eval.addEvalListener(new Eval2.EvalListener()
+        {
+            public void finishedEval(Eval2.EvalEvent aEvent)
+            {
+                synchronized (EvalTrace.this)
+                {
+                    try
+                    {
+                        // The eval is done. Set the flags.
+                        halted = true;
+                        hasResult = true;
+                        excepted = false;
+                        // Make data available.
+                        result = aEvent.getResult();
+                        stack = aEvent.getStack();
+                        exception = null;
+                        // Awake the debugger.
+                        EvalTrace.this.notifyAll();
+                        // The eval must wait, otherwise it might attempt
+                        // to evaluate a brand new expression which will destroy the result of
+                        // this earlier evaluation!
+                        EvalTrace.this.wait();
+                        halted = false;
+                    }
+                    catch (InterruptedException e)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)